mpmath/libmp/libmpf.py

```diff
@@ -1310,6 +1348,333 @@ def from_str(x, prec=0, rnd=round_fast, base=0):
+
+    return sign, digits
+
+
+def format_scientific(s,
+                      precision=6,
+                      strip_zeros=False,
+                      sign_spec='-',
+                      base=10,
+                      capitalize=False,
+                      alternate=False):
+
+    sep = 'E' if capitalize else 'e'
+
+    # First, get the exponent to know how many digits we will need
+    dps = precision+1
+    sign, digits, exponent = to_digits_exp(
+            s, max(dps+1, int(s[3]/blog2_10)), base)
+
+    if sign != '-' and sign_spec != '-':
+        sign = sign_spec
+
+    # Rounding up kills some instances of "...99999"
+    digits, exp_add = round_digits(digits, dps, base)
+    exponent += exp_add
+
+    if strip_zeros:
+        # Clean up trailing zeros
+        digits = digits.rstrip('0')
+        precision = len(digits)
+
+    if precision >= 1 and len(digits) > 1:
+        return sign, digits[0] + '.' + digits[1:] + sep + f'{exponent:+03d}'
+    else:
+        if alternate:
+            return sign, digits + '.' + sep + f'{exponent:+03d}'
+        else:
+            return sign, digits + sep + f'{exponent:+03d}'
+
+
+def format_zero(fmt_type, precision, alternate, capitalize,
+                thousands_separators):
+    digits = ''
+
+    if fmt_type in 'ef':
+        digits = '0'
+        # This is not used since fnzero is not used either, AFAIK.
+        # if num[0]:
+        #    sign = '-'
+
+        if precision > 0:
+            digits += '.' + precision*'0'
+        else:
+            if alternate:
+                digits += '.'
+
+        if fmt_type == 'e':
+            digits += 'E+00' if capitalize else 'e+00'
+
+    else:  # fmt_type == 'g'
+        if alternate:
+            return format_zero('f', precision-1, alternate, capitalize,
+                               thousands_separators)
+        else:
+            return '0'
+
+    return digits
+
+
+def format_mpf(num, format_spec):
+    format_dict = read_format_spec(format_spec)
+
+    capitalize = False
+    if format_dict['type'] in 'FGE':
+        capitalize = True
+
+    fmt_type = format_dict['type'].lower()
+    precision = format_dict['precision']
+
+    digits = ''
+    sign = ''
+
+    # Special cases:
+    if num == fnan:
+        sign = '' if format_dict['sign'] == '-' else format_dict['sign']
+        digits = 'NAN' if capitalize else 'nan'
+    elif num == finf:
+        sign = '' if format_dict['sign'] == '-' else format_dict['sign']
+        digits = 'INF' if capitalize else 'inf'
+    elif num == fninf:
+        sign = '-'
+        digits = 'INF' if capitalize else 'inf'
+    elif num == fzero:
+        sign = '' if format_dict['sign'] == '-' else format_dict['sign']
+        digits = format_zero(
+                fmt_type=fmt_type,
+                precision=precision, capitalize=capitalize,
+                alternate=format_dict['alternate'],
+                thousands_separators=format_dict['thousands_separators'])
```

Comment:
```suggestion
    # Special cases:
    if num in (fnan, finf, fninf):
        return format(format_spec, float(num))
    elif num == fzero:
        sign = '' if format_dict['sign'] == '-' else format_dict['sign']
        digits = format_zero(
                fmt_type=fmt_type,
                precision=precision, capitalize=capitalize,
                alternate=format_dict['alternate'],
                thousands_separators=format_dict['thousands_separators'])
```
Probably you could just fallback to python code here.  Maybe for zeros too.

Reply:
```suggestion
    # Special cases:
    if num in (fnan, finf, fninf, fzero, fnzero):
        return format(to_float(num), format_spec)
```
Yeah, that should work for all special numbers.